beacon-chain/verification/data_column.go

```diff
@@ -0,0 +1,504 @@
+		}
+
+		// Extract the slot of the data column.
+		dataColumnSlot := dataColumn.Slot()
+
+		// Check if the data column slot is after the parent slot.
+		if parentSlot >= dataColumnSlot {
+			return ErrSlotNotAfterParent
+		}
+	}
+
+	return nil
+}
+
+func (dv *RODataColumnsVerifier) SidecarDescendsFromFinalized() (err error) {
+	if ok, err := dv.results.cached(RequireSidecarDescendsFromFinalized); ok {
+		return err
+	}
+
+	defer dv.recordResult(RequireSidecarDescendsFromFinalized, &err)
+
+	for _, dataColumn := range dv.dataColumns {
+		// Extract the root of the parent block corresponding to the data column.
+		parentRoot := dataColumn.ParentRoot()
+
+		if !dv.fc.HasNode(parentRoot) {
+			return columnErrBuilder(ErrSidecarNotFinalizedDescendent)
+		}
+	}
+
+	return nil
+}
+
+func (dv *RODataColumnsVerifier) SidecarInclusionProven() (err error) {
+	if ok, err := dv.results.cached(RequireSidecarInclusionProven); ok {
+		return err
+	}
+
+	defer dv.recordResult(RequireSidecarInclusionProven, &err)
+
+	startTime := time.Now()
+
+	for _, dataColumn := range dv.dataColumns {
+		k, keyErr := inclusionProofKey(dataColumn)
+		if keyErr == nil {
+			if _, ok := dv.ic.Get(k); ok {
+				continue
+			}
+		} else {
+			log.WithError(keyErr).Error("Failed to get inclusion proof key")
+		}
+
+		if err = peerdas.VerifyDataColumnSidecarInclusionProof(dataColumn); err != nil {
+			return columnErrBuilder(ErrSidecarInclusionProofInvalid)
+		}
+
+		if keyErr == nil {
+			dv.ic.Add(k, struct{}{})
+		}
+	}
+
+	dataColumnSidecarInclusionProofVerificationHistogram.Observe(float64(time.Since(startTime).Milliseconds()))
+
+	return nil
+}
+
+func (dv *RODataColumnsVerifier) SidecarKzgProofVerified() (err error) {
+	if ok, err := dv.results.cached(RequireSidecarKzgProofVerified); ok {
+		return err
+	}
+
+	defer dv.recordResult(RequireSidecarKzgProofVerified, &err)
+
+	startTime := time.Now()
+
+	err = dv.verifyDataColumnsCommitment(dv.dataColumns)
+	if err != nil {
+		return columnErrBuilder(errors.Wrap(err, "verify data column commitment"))
+	}
+
+	dataColumnBatchKZGVerificationHistogram.Observe(float64(time.Since(startTime).Milliseconds()))
+	return nil
+}
+
+func (dv *RODataColumnsVerifier) SidecarProposerExpected(ctx context.Context) (err error) {
+	if ok, err := dv.results.cached(RequireSidecarProposerExpected); ok {
+		return err
+	}
+
+	defer dv.recordResult(RequireSidecarProposerExpected, &err)
+
+	for _, dataColumn := range dv.dataColumns {
+		// Extract the slot of the data column.
+		dataColumnSlot := dataColumn.Slot()
+
+		// Compute the epoch of the data column slot.
+		dataColumnEpoch := slots.ToEpoch(dataColumnSlot)
+		if dataColumnEpoch > 0 {
+			dataColumnEpoch = dataColumnEpoch - 1
+		}
+
+		// Extract the root of the parent block corresponding to the data column.
+		parentRoot := dataColumn.ParentRoot()
+
+		// Compute the target root for the epoch.
+		targetRoot, err := dv.fc.TargetRootForEpoch(parentRoot, dataColumnEpoch)
```

Comment:
Shouldn't the root and epoch be the same for every value in `dataColumns`? If so it would be way more concise, and efficient if any of these calls are expensive but uncached, to just check that all have matching root and slot, then perform the check once for all of them.

If they can be different, the same principle should apply, but first make a `map[root]slot`, and short-circuit if there is any conflict between roots and slots (same root different slot or same slot different roots).

Reply:
They can be different. But agree with your `map[root]slot` approach.

Reply:
Proposal in 496bea266e4e8beeed775428da4899e74b35d756.